### PR TITLE
sdr_trc_fix:

### PR DESCRIPTION
**Pull request by dylanraga**
```
- compute normalized value so that blacks are not raised by bt1886
- apply bt1886 eotf at input signal instead of its inverse at regamma
```

Currently, raising the BT.1886 black point also raises the black luminance on regamma. Normalizing the luminance fixes this. This way, the curve will look correct when viewed on a monitor with a non-zero black level.

Also, the BT.1886 eotf can be applied to only the input signal instead, and normally re-gamma'd rather than having two separate functions.

### CODE REVIEW

File: Shaders/lilium__sdr_trc_fix.fx [hunk 1]
```diff
@@ -69,38 +69,22 @@ uniform float BT1886_TARGET_BLACKPOINT
...
-  static const float powLw = pow(BT1886_TARGET_WHITEPOINT, TargetInverseGamma);
-  static const float powLb = pow(BT1886_TARGET_BLACKPOINT, TargetInverseGamma);
+  static const float targetInverseGamma = 1.f / gamma;
+  static const float powLw = pow(BT1886_TARGET_WHITEPOINT, targetInverseGamma);
+  static const float powLb = pow(BT1886_TARGET_BLACKPOINT, targetInverseGamma);
 
   static const float powLw_minus_powLb = powLw - powLb;
 
-  static const float a = pow(powLw_minus_powLb, TARGET_POWER_GAMMA);
+  static const float a = pow(powLw_minus_powLb, gamma);
```
Comment (EndlesslyFlowering) [inline Shaders/lilium__sdr_trc_fix.fx:80]: you don't need to feed uniform variables to functions in ReShade fx code. they are always accessible. I know that it is not like that hlsl.

**Comment by dylanraga** [comment]
Looks almost good here! The default option being switched is also a good touch. Just need to normalize the BT1886 value so that black luminance isn't raised,

ie
```
  float3 L_norm = (L - BT1886_TARGET_BLACKPOINT) / (BT1886_TARGET_WHITEPOINT - BT1886_TARGET_BLACKPOINT);

  return L_norm;
```

otherwise black levels are lifted on monitors with non-zero black luminance.

**Comment by EndlesslyFlowering** [comment]
> Looks almost good here! The default option being switched is also a good touch. Just need to normalize the BT1886 value so that black luminance isn't raised,
> 
> ie
> 
> ```
>   float3 L_norm = (L - BT1886_TARGET_BLACKPOINT) / (BT1886_TARGET_WHITEPOINT - BT1886_TARGET_BLACKPOINT);
> 
>   return L_norm;
> ```
> 
> otherwise black levels are lifted on monitors with non-zero black luminance.

Uhh that is the whole point of BT.1886 that the black level will be raised to whatever the display is capable of.
So if a black point of a display is 0.001 nits you want black (0) raised to 0.001 so that content black matches display black.
I know that it doesn't look that great but that's just how BT.1886 works. It affects the whole curve.

Alternatively I could implement the black point adjustment from the BT.2390/BT.2408 EETF. That is what I use in my shader `hdr_black_floor_fix`:
https://github.com/EndlesslyFlowering/ReShade_HDR_shaders/blob/6cdc04968efa1dcb0b76e8d6f1973e3f20e706a0/Shaders/lilium__include/hdr_black_floor_fix.fxh#L369-L563

**Comment by dylanraga** [comment]
The point of BT.1886 is to create a tailored Gamma 2.4-like curve on a display with a non-zero black level. For example, let's  imagine a typical IPS display with a white level of 100 nits and a black level of 0.01 nits (1000:1 contrast).

When outputting the current non-normalized BT.1886 fitted curve on this display, then black (#000000) is no longer 0.01 nits, but ~0.02 nits because the shader black point is being added on top of the display's existing black point. Right now, the shader models what the BT.1886 curve would look like for a certain white/black level, if viewed on a reference OLED. I'm assuming you want this shader to output the correct curve when viewed on the actual display for which the shader white-level and black-level are specified.

**Comment by EndlesslyFlowering** [comment]
> The point of BT.1886 is to create a tailored Gamma 2.4-like curve on a display with a non-zero black level. For example, let's imagine a typical IPS display with a white level of 100 nits and a black level of 0.1 nits (1000:1 contrast).
> 
> When outputting the current non-normalized BT.1886 fitted curve on this display, then black (#000000) is no longer 0.1 nits, but ~0.2 nits because the shader black point is being added on top of the display's existing black point. Right now, the shader models what the BT.1886 curve would look like for a certain white/black level, if viewed on a reference OLED. I'm assuming you want this shader to output the correct curve when viewed on the actual display for which the shader white-level and black-level are specified.
> 
> Perhaps add a toggle for “black point compensation”.

you are right

**Comment by EndlesslyFlowering** [comment]
implemented here:
eb3170463e6264ffd2538fe916f42c0ca1c6fcc3

**Pull request closed without merging**